File: packages/oxlint-plugin-react-doctor/src/plugin/utils/get-static-string-expression.ts
import type { EsTreeNode } from "./es-tree-node.js";
import { isNodeOfType } from "./is-node-of-type.js";

export const getStaticStringExpression = (
  expression: EsTreeNode | null | undefined,
): string | null => {
  if (isNodeOfType(expression, "Literal") && typeof expression.value === "string") {
    return expression.value;
  }
  if (isNodeOfType(expression, "TemplateLiteral") && expression.expressions.length === 0) {
    return expression.quasis[0]?.value.cooked ?? expression.quasis[0]?.value.raw ?? null;
  }
  return null;
};

File: native/oxlint/rules/get-static-string-expression.rs
fn get_static_string_expression<'a>(
    expression: &'a oxc_ast::ast::Expression<'a>,
) -> Option<&'a str> {
    match expression.get_inner_expression() {
        oxc_ast::ast::Expression::StringLiteral(string_literal) => {
            Some(string_literal.value.as_str())
        }
        oxc_ast::ast::Expression::TemplateLiteral(template_literal)
            if template_literal.expressions.is_empty() && template_literal.quasis.len() == 1 =>
        {
            let quasi = &template_literal.quasis[0];
            Some(
                quasi
                    .value
                    .cooked
                    .as_ref()
                    .map_or(quasi.value.raw.as_str(), |cooked| cooked.as_str()),
            )
        }
        _ => None,
    }
}
